File: product/src/application/service/product_service.py
from fastapi import HTTPException

from src.domain.interfaces.product_repository import AbstractProductRepository
from src.domain.models import Product
from src.infrastructure.db.models.product import ProductORM
from src.infrastructure.db.repositories.product import ProductRepository
from src.schemas.schemas import ProductCreate


class ProductService:
    def __init__(self,
                 product_repo: AbstractProductRepository):
        self.product_repo = product_repo

    def read_products(self,
                      skip: int = 0,
                      limit: int = 100,
                      ):
        products = self.product_repo.get_all_products(skip, limit)

        return products

    def create_product(self,
                       product: ProductCreate
                       ) -> None | Product:
        product_exists = self.product_repo.check_if_exists(product.name)
        if product_exists:
            raise HTTPException(status_code=400, detail="Product with this name already exists")
        else:
            new_product = Product(
                name=product.name,
                short_description=product.short_description,
                full_description=product.full_description,
                composition=product.composition,
                weight=product.weight,
                price=product.price,
                photo=product.photo
            )

            return self.product_repo.add_product(new_product)

    def update_product(self,
                       id: int,
                       product: ProductCreate
                       ):
        product_exists = self.product_repo.check_if_exists(product.name)
        if product_exists:
            raise HTTPException(status_code=400, detail="Product with this name already exists")
        else:
            self.product_repo.save(product)

    def verify_product(self,
                       name: str
                       ):
        product_exists = self.product_repo.get_by_name(name)
        if product_exists:
            return {"exists": True,
                    "id": product_exists.id,
                    "name": product_exists.name}
        else:
            return {"exists": False,
                    "id": None,
                    "name": None}

    def info_product(self,
                     id: int
                     ):
        product_exists = self.product_repo.get_product_by_id(id)
        if product_exists:
            return {"product": product_exists}
        else:
            raise HTTPException(status_code=404, detail="Product not found")

File: data/internal/application/service/product_service.go
package service

import (
	"context"
	"data-service/internal/application/dto"
	"data-service/internal/domain/entity"
	"data-service/internal/domain/repository"
	"errors"
	"time"
)

var (
	ErrProductNotFound = errors.New("product not found")
)

// ProductService сервис для работы с продуктами
type ProductService struct {
	repo repository.ProductRepository
}

// NewProductService создает новый экземпляр ProductService
func NewProductService(repo repository.ProductRepository) *ProductService {
	return &ProductService{repo: repo}
}

// CreateProduct создает новый продукт
func (s *ProductService) CreateProduct(ctx context.Context, req *dto.CreateProductRequest) (*dto.ProductResponse, error) {
	product := &entity.Product{
		Name:             req.Name,
		ShortDescription: req.ShortDescription,
		FullDescription:  req.FullDescription,
		Composition:      req.Composition,
		Weight:           req.Weight,
		Price:            req.Price,
		Photo:            req.Photo,
		CreatedAt:        time.Now(),
		UpdatedAt:        time.Now(),
	}

	createdProduct, err := s.repo.Create(ctx, product)
	if err != nil {
		return nil, err
	}

	return s.toProductResponse(createdProduct), nil
}

// GetProductByID получает продукт по ID
func (s *ProductService) GetProductByID(ctx context.Context, id int) (*dto.ProductResponse, error) {
	product, err := s.repo.GetByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if product == nil {
		return nil, ErrProductNotFound
	}

	return s.toProductResponse(product), nil
}

// GetProductByName получает продукт по названию
func (s *ProductService) GetProductByName(ctx context.Context, name string) (*dto.ProductResponse, error) {
	product, err := s.repo.GetByName(ctx, name)
	if err != nil {
		return nil, err
	}
	if product == nil {
		return nil, ErrProductNotFound
	}

	return s.toProductResponse(product), nil
}

// ListProducts получает список всех продуктов
func (s *ProductService) ListProducts(ctx context.Context, filter *entity.ProductFilter) ([]*dto.ProductResponse, error) {
	products, err := s.repo.List(ctx, filter)
	if err != nil {
		return nil, err
	}

	responses := make([]*dto.ProductResponse, 0, len(products))
	for _, product := range products {
		responses = append(responses, s.toProductResponse(product))
	}

	return responses, nil
}

// UpdateProduct обновляет данные продукта
func (s *ProductService) UpdateProduct(ctx context.Context, id int, req *dto.UpdateProductRequest) (*dto.ProductResponse, error) {
	product, err := s.repo.GetByID(ctx, id)
	if err != nil {
		return nil, err
	}
	if product == nil {
		return nil, ErrProductNotFound
	}

	// Обновляем только переданные поля
	if req.Name != nil {
		product.Name = *req.Name
	}
	if req.ShortDescription != nil {
		product.ShortDescription = *req.ShortDescription
	}
	if req.FullDescription != nil {
		product.FullDescription = *req.FullDescription
	}
	if req.Composition != nil {
		product.Composition = *req.Composition
	}
	if req.Weight != nil {
		product.Weight = *req.Weight
	}
	if req.Price != nil {
		product.Price = *req.Price
	}
	if req.Photo != nil {
		product.Photo = *req.Photo
	}
	product.UpdatedAt = time.Now()

	updatedProduct, err := s.repo.Update(ctx, product)
	if err != nil {
		return nil, err
	}

	return s.toProductResponse(updatedProduct), nil
}

// DeleteProduct удаляет продукт по ID
func (s *ProductService) DeleteProduct(ctx context.Context, id int) error {
	return s.repo.Delete(ctx, id)
}

// CheckProductExists проверяет существование продукта
func (s *ProductService) CheckProductExists(ctx context.Context, id int) (bool, error) {
	return s.repo.Exists(ctx, id)
}

// toProductResponse преобразует entity.Product в dto.ProductResponse
func (s *ProductService) toProductResponse(product *entity.Product) *dto.ProductResponse {
	return &dto.ProductResponse{
		ID:               product.ID,
		Name:             product.Name,
		ShortDescription: product.ShortDescription,
		FullDescription:  product.FullDescription,
		Composition:      product.Composition,
		Weight:           product.Weight,
		Price:            product.Price,
		Photo:            product.Photo,
		CreatedAt:        product.CreatedAt.Format(time.RFC3339),
		UpdatedAt:        product.UpdatedAt.Format(time.RFC3339),
	}
}

